oap-ape/ape-java/ape-server/src/main/java/com/intel/oap/ape/service/netty/server/RequestHandler.java

```diff
@@ -184,24 +195,30 @@ private void handleParquetReaderInitRequest(NettyMessage.ParquetReaderInitReques
         }
     }
 
-    private NettyMessage.ReadBatchResponse handleReadBatchRequest(ChannelHandlerContext ctx) {
+    private NettyMessage.ReadBatchResponse readNextBatch(Channel channel) {
         final long start = System.nanoTime();
 
+        int rowCount = 0;
+        hasNextBatch = false;
+
         // read batch
-        int rowCount = ParquetReaderJNI.readBatch(
-                reader, batchSize, nativeDataBuffers, nativeNullBuffers);
-        if (rowCount < 0) {
-            rowCount = 0;
-        }
+        if (rowGroupsToRead > 0) {
+            rowCount = ParquetReaderJNI.readBatch(
+                    reader, batchSize, nativeDataBuffers, nativeNullBuffers);
+
+            if (rowCount < 0) {
+                rowCount = 0;
+            }
 
-        // check next batch
-        boolean hasNextBatch = ParquetReaderJNI.hasNext(reader);
+            // check next batch
+            hasNextBatch = ParquetReaderJNI.hasNext(reader);
+        }
 
         // header info
         boolean[] compositeFlags = new boolean[columnCount];
         int[] dataBufferLengths = new int[columnCount];
         int compositedElementCount = 0;
-        ByteBuf compositedElementLengths = ctx.alloc().ioBuffer(4 * rowCount);
+        ByteBuf compositedElementLengths = channel.alloc().ioBuffer(4 * rowCount);

```

Comment:
why 4 here? try to avoid magic number?

Reply:
It's to set the initial capacity of a buffer. This buffer is used to store lengths of byte arrays. So all elements will be integers.
`4 * rowCount` means that the initial capacity is a length that can hold 1 column of elements.

Reply:
Got it. Thanks.